client-server-client/src/main/scala/zio/raft/client/RaftClient.scala

```diff
@@ -0,0 +1,774 @@
+        */
+      private def reconnectTo(
+        targetMemberId: Option[MemberId],
+        transport: ZmqClientTransport,
+        logMessage: String
+      ): UIO[ConnectingExistingSession] = {
+        val (memberId, address) = targetMemberId
+          .flatMap(id => clusterMembers.get(id).map(addr => (id, addr)))
+          .getOrElse(nextMember)
+
+        for {
+          _ <- ZIO.logInfo(logMessage)
+          nonce <- Nonce.generate()
+          _ <- transport.disconnect().orDie
+          _ <- transport.connect(address).orDie
+          _ <- transport.sendMessage(ContinueSession(sessionId, nonce)).orDie
+          _ <- ZIO.logInfo(s"Connecting Existing Session $sessionId to $memberId at $address")
+          now <- Clock.instant
+        } yield ConnectingExistingSession(
+          sessionId = sessionId,
+          capabilities = capabilities,
+          nonce = nonce,
+          clusterMembers = clusterMembers,
+          currentMemberId = memberId,
+          createdAt = now,
+          serverRequestTracker = serverRequestTracker,
+          nextRequestId = nextRequestId,
+          pendingRequests = pendingRequests
+        )
+      }
+
+      override def handle(
+        event: StreamEvent,
+        transport: ZmqClientTransport,
+        config: ClientConfig,
+        serverRequestQueue: Queue[ServerRequest]
+      ): UIO[ClientState] = {
+        event match {
+          case StreamEvent.Action(ClientAction.Disconnect) =>
+            for {
+              _ <- transport.sendMessage(CloseSession(CloseReason.ClientShutdown)).orDie
+              _ <- transport.disconnect().orDie
+              _ <- ZIO.logInfo("Disconnected")
+            } yield Disconnected
+
+          case StreamEvent.Action(ClientAction.Connect) =>
+            ZIO.logWarning("Received Connect while connected, ignoring").as(this)
+
+          case StreamEvent.Action(ClientAction.SubmitCommand(payload, promise)) =>
+            for {
+              requestId <- nextRequestId.next
+              now <- Clock.instant
+              request = ClientRequest(requestId, payload, now)
+              _ <- transport.sendMessage(request).orDie
+              newPending = pendingRequests.add(requestId, payload, promise, now)
+            } yield copy(pendingRequests = newPending)
+
+          case StreamEvent.ServerMsg(ClientResponse(requestId, result)) =>
+            pendingRequests.complete(requestId, result).map(newPending => copy(pendingRequests = newPending))
+
+          case StreamEvent.ServerMsg(KeepAliveResponse(_)) =>
+            ZIO.succeed(this)
+
+          case StreamEvent.ServerMsg(serverRequest: ServerRequest) =>
+            serverRequestTracker.shouldProcess(serverRequest.requestId) match {
+              case ServerRequestResult.Process =>
+                for {
+                  _ <- serverRequestQueue.offer(serverRequest)
+                  newTracker = serverRequestTracker.acknowledge(serverRequest.requestId)
```

Comment:
do we want to auto ack or let the user let us know when it is safe to ack?

Reply:
auto ack is fine here, all we care is that it's passed the network. If the user didn't handle it means the pod crashed, which means the session is died. So auto-acking is fine.